cli/lsp/tsc.rs

```diff
@@ -4576,9 +4572,92 @@ impl State {
   }
 }
 
+#[op2]
+#[serde]
+fn op_export_modules_for_module(
+  state: &mut OpState,
+  #[string] module: String,
+) -> IndexSet<String> {
+  let Ok(referrer) = ModuleSpecifier::parse(&module) else {
+    return Default::default();
+  };
+  let state = state.borrow::<State>();
+  let scope_resolver = state
+    .state_snapshot
+    .resolver
+    .get_scope_resolver(Some(&referrer));
+  let config_data = scope_resolver.as_config_data();
+  let mut urls = IndexSet::new();
+
+  get_jsr_and_npm_importable_paths(scope_resolver, &referrer, state, &mut urls);
+
+  // Include all remote modules and any modules in the current
+  // package (or else all user code)
+  let documents = state
+    .state_snapshot
+    .documents
+    .documents(DocumentsFilter::AllDiagnosable);
+  let in_npm_pkg_checker = scope_resolver.as_in_npm_pkg_checker();
+  for document in documents {
+    let should_add = match document.specifier().scheme() {
+      // todo: also filter out specifiers in other packages
+      "file" => !in_npm_pkg_checker.in_npm_package(document.specifier()),
+      "cache" => false,
+      _ => true,
+    };
+    if should_add {
+      urls.insert(state.specifier_map.denormalize(document.specifier()));
+    }
+  }
+
+  urls
+}
+
+fn get_jsr_and_npm_importable_paths(
+  scope_resolver: &LspScopeResolver,
+  referrer: &ModuleSpecifier,
+  state: &State,
+  urls: &mut IndexSet<String>,
+) {
+  // Discover all the npm package export modules
+  if let Some(npm_resolver) = scope_resolver.as_npm_resolver() {
+    if let Some(node_resolver) = scope_resolver.as_node_resolver() {
+      let dep_info = scope_resolver.dep_info();
+      for req in &dep_info.npm_reqs {
+        let maybe_package_folder = npm_resolver
+          .resolve_pkg_folder_from_deno_module_req(&req, referrer)
+          .ok();
+        if let Some(pkg_folder) = maybe_package_folder {
+          let exports = node_resolver.all_exported_modules_for_package(
+            &pkg_folder,
+            // todo: check if the referrer is cjs and provide require
+            ResolutionMode::Import,
+            NodeResolutionKind::Types,
+          );
```

Comment:
We'll need to add some caching for this because it will be slow for certain packages.